sw/device/tests/alert_handler_ping_ok_test.c

```diff
@@ -0,0 +1,245 @@
+  CHECK_DIF_OK(dif_rv_timer_counter_read(&timer, kHart,
+                                         (uint64_t *)&counters.elapsed_ticks));
+  CHECK(counters.elapsed_ticks == 0, "Failed to write the counter");
+
+  CHECK_DIF_OK(
+      dif_rv_timer_counter_set_enabled(&timer, kHart, kDifToggleEnabled));
+  CHECK_DIF_OK(dif_alert_handler_configure_ping_timer(
+      &alert_handler, config.ping_timeout, kDifToggleEnabled,
+      kDifToggleEnabled));
+}
+
+/**
+ * Execute the alert handler NMI interrupt test.
+ */
+static void alert_handler_ping_ok_test(void) {
+  alert_handler_config();
+
+  uint64_t theshold_us = kTimeThresholdUs;
+  if (kDeviceType != kDeviceSimDV && kDeviceType != kDeviceSimVerilator) {
+    theshold_us = theshold_us * 10;
+    CHECK(theshold_us > kTimeThresholdUs, "Threshold overflow");
+  }
+
+  do {
+    CHECK_DIF_OK(dif_rv_timer_counter_read(
+        &timer, kHart, (uint64_t *)&counters.elapsed_ticks));
+  } while (counters.elapsed_ticks < theshold_us);
+
+  CHECK_DIF_OK(
+      dif_rv_timer_counter_set_enabled(&timer, kHart, kDifToggleDisabled));
+}
+
+/**
+ * OTTF external NMI internal IRQ handler.
+ * This functions overrides the OTTF weak definition.
+ */
+void ottf_external_nmi_handler(uint32_t *exc_info) {
+  CHECK_DIF_OK(dif_rv_core_ibex_get_nmi_state(
+      &rv_core_ibex, (dif_rv_core_ibex_nmi_state_t *)&nmi_state));
+
+  if (nmi_state.wdog_barked) {
+    counters.wdog_barked++;
+  }
+  if (nmi_state.alert_raised) {
+    counters.alert_raised++;
+  }
+
+  CHECK_DIF_OK(dif_rv_core_ibex_clear_nmi_state(&rv_core_ibex,
+                                                kDifRvCoreIbexNmiSourceAll));
+}
+
+/**
+ * OTTF external IRQ handler
+ * This functions overrides the OTTF weak definition.
+ */
+void ottf_external_isr(uint32_t *exc_info) { ext_irq_fired = true; }
+
+void init_peripherals(void) {
+  mmio_region_t addr =
+      mmio_region_from_addr(TOP_EARLGREY_RV_CORE_IBEX_CFG_BASE_ADDR);
+  CHECK_DIF_OK(dif_rv_core_ibex_init(addr, &rv_core_ibex));
+
+  addr = mmio_region_from_addr(TOP_EARLGREY_ALERT_HANDLER_BASE_ADDR);
+  CHECK_DIF_OK(dif_alert_handler_init(addr, &alert_handler));
+
+  addr = mmio_region_from_addr(TOP_EARLGREY_PWRMGR_AON_BASE_ADDR);
+  CHECK_DIF_OK(dif_pwrmgr_init(addr, &pwrmgr));
+
+  CHECK_DIF_OK(dif_rv_timer_init(
+      mmio_region_from_addr(TOP_EARLGREY_RV_TIMER_BASE_ADDR), &timer));
+  CHECK_DIF_OK(dif_rv_timer_reset(&timer));
+}
+
+bool test_main(void) {
+  // Disable external interrupts via the PLIC entirely.
+  // The NMI should still get through to the processor regardless.
+  irq_global_ctrl(false);
+  irq_external_ctrl(false);
+  init_peripherals();
+
+  CHECK_DIF_OK(
+      dif_rv_core_ibex_enable_nmi(&rv_core_ibex, kDifRvCoreIbexNmiSourceAll));
+
+  alert_handler_ping_ok_test();
+
+  // Double check that no external interrupts have occurred.
+  CHECK(ext_irq_fired == false, "Unexpected external interrupt triggered.");
```

Comment:
Do you also need to check that no NMIs have fired? (`counters.alert_raised`)

Reply:
Good catch. Thanks! Fixed.